src/lib/Syntax.hs

```diff
@@ -103,6 +104,10 @@ data Atom = Var Var
             -- single-constructor only for now
           | DataConRef DataDef [Atom] (Nest DataConRefBinding)
           | BoxedRef Binder Atom Block Atom  -- binder, ptr, size, body
+          -- access a nested member of a binder
+          -- XXX: Variable name MUST be fully reduced, it cannot be a synonym!
+          -- This is because the variable name may also appear in the type.
+          | ProjectElt (NE.NonEmpty Int) Var
```

Comment:
Yeah, I'm not sure exactly what the right way to phrase this is. Variable name must be lambda-bound, not let-bound? Variable name must not be an alias for a known value?

`ProjectElt Int Var` doesn't work, we would have to have `ProjectElt Int Atom` (maybe that's what you meant?). I'd be happy to rewrite it in terms of `ProjectElt Int Atom`, but the downside there is that we no longer guarantee in the Haskell ADT that the `Atom` can only be a `ProjectElt` or a `Var` (e.g. we could build a `ProjectElt 0 (PairVal x y)` which should never actually exist in the IR because it could be simplified). Perhaps we have enough invariant checking that this is fine, although it still has the danger of turning a Haskell type error into a runtime error.

(Regarding your top-level comment, `ProjectElt [0] (ProjectElt [1] something))` can't possibly exist right now, because `ProjectElt [1] something` isn't a `Var`. So the only possible representation of this as an atom is `ProjectElt [0, 1] something`.)

Reply:
Yeah, binding sites sound like a good way to distinguish them, but I don't think it is accurate. Not sure if we have any `PrimOp`s like that, but if any of them returned records or ADTs, then those variables could also be let-bound.

Right, good point about the recursion being necessary, because it has to terminate in `Var` and not in an `Atom`. I'm convinced now, thanks for the explanation.

Reply:
Yeah, you're right, it's not accurate to say it's based on binding sites. I'll clarify the comment to remove the "fully reduced" part and replace it with something about aliases.